osu.Game/Rulesets/Mods/ModHealthChallenge.cs

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Bindables;
+using osu.Game.Configuration;
+using osu.Game.Rulesets.Judgements;
+using osu.Game.Rulesets.Scoring;
+
+namespace osu.Game.Rulesets.Mods
+{
+    public class ModHealthChallenge : ModChallenge, IApplicableToHealthProcessor
+    {
+        public override string Name => "Health Challenge";
+        public override string Acronym => "HC";
+        public override string Description => "Hold your health or you fail";
+
+        [SettingSource("Health Threshold", "Set your own health threshold")]
+        public Bindable<double> HealthThreshold { get; } = new BindableDouble
+        {
+            Precision = 0.01,
+            MinValue = 0.50,
+            MaxValue = 1.00,
+            Default = 0.96,
+            Value = 0.96,
+        };
+
+        public override string SettingDescription => $"{HealthThreshold.Value:0%}";
+
+        public void ApplyToHealthProcessor(HealthProcessor healthProcessor)
+        {
+            healthProcessor.FailConditions += checkThresholdCondition;
+        }
+
+        private bool checkThresholdCondition(HealthProcessor healthProcessor, JudgementResult result)
+        {
+            // Need to manually checked here since HasCompleted is updated after ApplyResultInternal
+            bool allJudged = healthProcessor.JudgedHits == healthProcessor.MaxHits;
```

Comment:
this _is_ the correct way to check for map completion, but as far as I recall this is already done the same way in other places and as such I think it should be exposed in a better way than just making `MaxHits` internal.

also not sure whether this health threshold should be checked only at the end instead of being enforced all the time. not sure which one is more playable.

Reply:
It's actually same logic that `JudgementProcessor` use to check beatmap completion and exposed via `HasCompleted` bindable. But since `HasCompleted` [updated after](https://github.com/ppy/osu/blob/master/osu.Game/Rulesets/Scoring/JudgementProcessor.cs#L55-#L61) `ApplyResultInternal` it basically no use here. I'm not sure if there's another exposed variable to check beatmap completion anywhere else.

Also I guess it's a good idea to make threshold checking toggle-able so player can switch between beatmap completion and during gameplay. Will try to make it toggle-able and let's see the feedback.

Reply:
I don't mind the way it's done here.

Reply:
I am aware that it's the same logic. I'm just saying it should maybe be exposed better overall (I had a need for that same logic once before and just opted not to bother there). `IsLastJudgement` as a property, maybe? Although maybe that's not too good of a property to have public... Not sure.